rust/lance/src/datafusion/logical_plan.rs

```diff
@@ -0,0 +1,144 @@
+use std::{any::Any, sync::Arc};
+
+use arrow_schema::Schema as ArrowSchema;
+use async_trait::async_trait;
+use datafusion::{
+    datasource::TableProvider,
+    error::Result as DatafusionResult,
+    execution::context::SessionState,
+    logical_expr::{LogicalPlan, TableProviderFilterPushDown, TableType},
+    optimizer::utils::conjunction,
+    physical_plan::{filter::FilterExec, limit::GlobalLimitExec, ExecutionPlan, Statistics},
+    prelude::Expr,
+};
+
+use crate::{
+    datafusion::physical_expr::column_names_in_expr,
+    io::exec::{Planner, ProjectionExec},
+    Dataset,
+};
+
+#[async_trait]
+impl TableProvider for Dataset {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> Arc<ArrowSchema> {
+        Arc::new(self.schema().into())
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    fn get_table_definition(&self) -> Option<&str> {
+        None
+    }
+
+    fn get_logical_plan(&self) -> Option<&LogicalPlan> {
+        None
+    }
+
+    async fn scan(
+        &self,
+        _: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
+        let scanner = self.scan();
+        let filter_expr = conjunction(filters.iter().cloned());
+        // NOTE: we only support node that have one partition. So any nodes that
+        // produce multiple need to be repartitioned to 1.
+        let mut filter_expr = if let Some(filter) = filter_expr {
+            let planner = Planner::new(Arc::new(scanner.dataset.schema().into()));
+            Some(planner.create_physical_expr(&filter)?)
+        } else {
+            None
+        };
+
+        // Stage 1: source
+        let mut plan: Arc<dyn ExecutionPlan> = if scanner.nearest.is_some() {
+            if scanner.prefilter {
+                let prefilter = filter_expr;
+                filter_expr = None;
+                scanner.knn(prefilter).await?
+            } else {
+                scanner.knn(None).await?
+            }
+        } else if let Some(expr) = filter_expr.as_ref() {
+            let columns_in_filter = column_names_in_expr(expr.as_ref());
+            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
+            scanner.scan(true, false, filter_schema)
+        } else {
+            // Scan without filter or limits
+            scanner.scan(
+                scanner.with_row_id,
+                false,
+                scanner.projections.clone().into(),
+            )
+        };
```

Comment:
I think we can ignore a bunch of options that aren't used:

```suggestion
        if let Some(expr) = filter_expr.as_ref() {
            let columns_in_filter = column_names_in_expr(expr.as_ref());
            let filter_schema = Arc::new(scanner.dataset.schema().project(&columns_in_filter)?);
            scanner.scan(true, false, filter_schema)
        } else {
            // Scan without filter or limits
            scanner.scan(
                false,
                false,
                scanner.projections.clone().into(),
            )
        };
```